### PR TITLE
resolve: reallocate DNS scope when DNSSEC and/or DNS-over-TLS settings are changed

### LINKED ISSUE #23227: Disabling DNSSEC per link does not work (systemd-resolved)
**systemd version the issue has been seen with**
 > 250.4-2

**Used distribution**
 > archlinux

**Linux kernel version used** (`uname -a`)

 > `5.17.5-zen1`

**CPU architecture issue was seen on**
 > x64

**Unexpected behaviour you saw**
 DNS validation failed on links without DNSSEC validation enabled.

**Steps to reproduce the problem**
 1. Using `DNSSEC=yes` as a default: everything works.
 2. Setting up a new link with a specific search domain (e.g. ~dummy.com) and another DNS IP
 3. Disable DNSSEC on the new link (DNS server is not compatible)
 4. `resolve query test.dummy.com`: `test.dummy.com: resolve call failed: DNSSEC validation failed: incompatible-server`

Note: this exact same process with DNS-over-TLS works as expected: when disabling DNS/TLS on a specific link, it does actually disable it.

When disabling DNSSEC entirely, everything works.

**Additional program output to the terminal or log subsystem illustrating the issue**

resolvectl output:

```text
Global
           Protocols: -LLMNR +mDNS +DNSOverTLS DNSSEC=yes/supported
    resolv.conf mode: stub
  Current DNS Server: 9.9.9.9#dns.quad9.net
         DNS Servers: 9.9.9.9#dns.quad9.net 149.112.112.112#dns.quad9.net 2620:fe::fe#dns.quad9.net 2620:fe::9#dns.quad9.net
Fallback DNS Servers: 1.1.1.1#cloudflare-dns.com 9.9.9.9#dns.quad9.net 8.8.8.8#dns.google 2606:4700:4700::1111#cloudflare-dns.com 2620:fe::9#dns.quad9.net
                      2001:4860:4860::8888#dns.google
          DNS Domain: ~.

Link 14 (geeTest)
    Current Scopes: DNS
         Protocols: +DefaultRoute +LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
Current DNS Server: 172.30.0.2
       DNS Servers: 172.30.0.2
        DNS Domain: ~. ~dummy.com

```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #23227.

### CODE REVIEW

File: src/resolve/resolved-link.c [hunk 1]
```diff
@@ -390,6 +390,9 @@ void link_set_dns_over_tls_mode(Link *l, DnsOverTlsMode mode) {
 #endif
 
         l->dns_over_tls_mode = mode;
+
+        if (l->unicast_scope)
```
Comment (daandemeyer) [inline src/resolve/resolved-link.c]: No need for the if check here I think, `dns_scope_free()` does its own check for `NULL` internally

File: src/resolve/resolved-link.c [hunk 2]
```diff
@@ -430,17 +433,10 @@ void link_set_dnssec_mode(Link *l, DnssecMode mode) {
...
-
-                /* When switching from non-DNSSEC mode to DNSSEC mode, flush the cache. Also when switching from the
-                 * allow-downgrade mode to full DNSSEC mode, flush it too. */
-                if (l->unicast_scope)
-                        dns_cache_flush(&l->unicast_scope->cache);
-        }
-
         l->dnssec_mode = mode;
+
+        if (l->unicast_scope)
```
Comment (daandemeyer) [inline src/resolve/resolved-link.c]: Same here

File: src/resolve/resolved-link-bus.c [hunk 3]
```diff
@@ -640,6 +640,7 @@ int bus_link_method_set_dns_over_tls(sd_bus_message *message, void *userdata, sd
 
         if (l->dns_over_tls_mode != mode) {
                 link_set_dns_over_tls_mode(l, mode);
+                link_allocate_scopes(l);
```
Comment (daandemeyer) [inline src/resolve/resolved-link-bus.c:643]: Why not call this in `link_set_dns_over_tls_mode()`?

File: src/resolve/resolved-link-bus.c [hunk 4]
```diff
@@ -688,6 +689,7 @@ int bus_link_method_set_dnssec(sd_bus_message *message, void *userdata, sd_bus_e
 
         if (l->dnssec_mode != mode) {
                 link_set_dnssec_mode(l, mode);
+                link_allocate_scopes(l);
```
Comment (daandemeyer) [inline src/resolve/resolved-link-bus.c:692]: Same here, why not call `link_allocate_scopes()` in `link_set_dnssec_mode()`?

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Some nits and a question

File: src/resolve/resolved-link-bus.c (hunk 3 above)
Comment (yuwata) [reply src/resolve/resolved-link-bus.c:643]: `link_set_dns_over_tls_mode()` is called on sd-network change, and in that case, `link_allocate_scope()` will be called after all sd-network properties are parsed.

**Review by DaanDeMeyer — APPROVED** [review]

**Comment by bluca** [comment]
Any chance test coverage for this issue can be added?

**Comment by yuwata** [comment]
> Any chance test coverage for this issue can be added?

I think it is better to add a testcase in #23104.

**Pull request merged**